### DISCUSSION TITLE
Heterogeneous graph model application design problem? GCN, GAT

### DISCUSSION THREAD

**Discussion by Zhang-1218** [pyg-team/pytorch_geometric#4224] [Q&A] [answered]
Heterogeneous graph model application design problem? GCN, GAT

Hello, I am learning graph network recently, thank you for providing the pyg open source library

My demo dataset is a heterogeneous graph data, referring to the dataset designed by the official website, which contains two types of nodes and two types of edges
Node:
(1) paper
(2) auther
Edge（edge is undirected）:
(1) paper-cite-paper
(2) auther-write-paper

![image](https://user-images.githubusercontent.com/28587429/157361581-dcbf9ea6-36ce-4354-bc74-0e7cba2ccfdf.png)

The data is as follows：
```
HeteroData(
  paper={
    x=[156, 7],
    y=[156],
    num_nodes=156,
    train_mask=[156],
    val_mask=[156],
    test_mask=[156]
  },
  author={
    x=[152, 7],
    y=[152],
    num_nodes=152
  },
  (author, write, paper)={
    edge_index=[2, 365],
    edge_attr=[365, 3]
  },
  (paper, cite, paper)={
    edge_index=[2, 2675],
    edge_attr=[2675, 6]
  },
  (paper, rev_write, author)={
    edge_index=[2, 365],
    edge_attr=[365, 3]
  }
)
```
Both author and paper nodes have features.
Both paper-cite-paper and auther-write-paper also have features
The purpose of the experiment is to hope to predict the category of the paper

1、I'm currently running into a problem:
```
  File "GCNProject/model/baseHeteModel.py", line 64, in basehetegat_train
    out = hgtmodel(data.x_dict, data.edge_index_dict)
  File "/root/anaconda3/envs/lib/python3.8/site-packages/torch/fx/graph_module.py", line 308, in wrapped_call
    return cls_call(self, *args, **kwargs)
  File "/root/anaconda3/envs/lib/python3.8/site-packages/torch/fx/graph_module.py", line 308, in wrapped_call
    return cls_call(self, *args, **kwargs)
  File "/root/anaconda3/envs/lib/python3.8/site-packages/torch/nn/modules/module.py", line 889, in _call_impl
    result = self.forward(*input, **kwargs)
  File "<eval_with_key_1>", line 8, in forward
  File "/root/anaconda3/envs/lib/python3.8/site-packages/torch/nn/modules/module.py", line 889, in _call_impl
    result = self.forward(*input, **kwargs)
  File "/root/anaconda3/envs/lib/python3.8/site-packages/torch_geometric/nn/conv/gat_conv.py", line 218, in forward
    out = self.propagate(edge_index, x=x, alpha=alpha, edge_attr=edge_attr,
  File "/root/anaconda3/envs/lib/python3.8/site-packages/torch_geometric/nn/conv/message_passing.py", line 262, in propagate
    size = self.__check_input__(edge_index, size)
  File "/root/anaconda3/envs/lib/python3.8/site-packages/torch_geometric/nn/conv/message_passing.py", line 149, in __check_input__
    raise ValueError(
ValueError: `MessagePassing.propagate` only supports `torch.LongTensor` of shape `[2, num_messages]` or `torch_sparse.SparseTensor` for argument `edge_index`.
```

I refer to the official website documentation and design the following model:
```
class HeteroGNN(torch.nn.Module):
    def __init__(self, hidden_channels, out_channels):
        super().__init__()
        self.conv1 = GATConv((-1, -1), hidden_channels, add_self_loops=False)
        self.lin1 = Linear(-1, hidden_channels)
        self.conv2 = GATConv((-1, -1), out_channels, add_self_loops=False)
        self.lin2 = Linear(-1, out_channels)

    def forward(self, x, edge_index):
        x = self.conv1(x, edge_index) + self.lin1(x)
        x = x.relu()
        x = self.conv2(x, edge_index) + self.lin2(x)
        return x


model = HeteroGNN(hidden_channels=4, out_channels=dataset.num_classes)
hgtmodel = to_hetero(model, data.metadata(), aggr='sum')

hgtmodel.train()
optimizer.zero_grad()
out = hgtmodel(data.x_dict, data.edge_index_dict)

```

I hope to be able to introduce both node and edge features into the calculation, but the documentation on the official website does not show how to use edge features for learning on heterogeneous graphs converted from homogeneous graphs.（GAT、GCN）
[https://pytorch-geometric.readthedocs.io/en/latest/notes/heterogeneous.html](url)

2、I saw on the official website that the heterogeneous convolution wrapper can be used to learn different models for different edges, so for my demo, is it possible to set paper-cite-paper to use gcn, auther-write-paper to use GAT, paper-rev_write-auther using GraphSAGE?
[https://pytorch-geometric.readthedocs.io/en/latest/notes/heterogeneous.html#using-the-heterogenous-convolution-wrapper](url)

If possible, my question is still how to pass edge features to convolutional models corresponding to different edges?

**Comment by rusty1s** [comment, marked as answer]
Can you upload the data (e.g., via `torch.save(data, 'data.pt')`? I'm happy to look into the issue.

You should be able to use any GNN operator that supports edge features as well, e.g.:
```python
conv = GATConv((-1, -1), hidden_channels, edge_dim=-1)
```
and call it via
```python
conv(x, edge_index, edge_attr)
```
Heterogeneous models can then be called via
```python
hgtmodel(data.x_dict, data.edge_index_dict, data.edge_attr_dict)
````

2. Yes, that is possible. The procedure to add edge features is the same (but you need to use GNN operators that can operate on them).

**Reply by Zhang-1218** [reply]
Thank you for your reply
I can store the processed data normally
```
torch.save(self.collate([data]), self.processed_paths[0])
```
There may be a problem with the processing of the data？
There are 1934 edges (2, 1934) in the original data (paper, cite, paper), but after T.ToUndirected()(data) processing, it becomes 2675 (2, 2675).
I don't understand how ToUndirected handles edges between nodes of the same kind

**Reply by rusty1s** [reply]
This looks correct to me. In case there already exists bidirectional edges between two nodes, `ToUndirected` will not add another edge.

**Reply by Zhang-1218** [reply]
Here is one of my test dataset

**Reply by rusty1s** [reply]
The following code runs through for me. Can you confirm?
```python
import torch

from torch_geometric.nn import GATConv, Linear, to_hetero

data = torch.load('data.pt')[0]


class HeteroGNN(torch.nn.Module):
    def __init__(self, hidden_channels, out_channels):
        super().__init__()
        self.conv1 = GATConv((-1, -1), hidden_channels, add_self_loops=False)
        self.lin1 = Linear(-1, hidden_channels)
        self.conv2 = GATConv((-1, -1), out_channels, add_self_loops=False)
        self.lin2 = Linear(-1, out_channels)

    def forward(self, x, edge_index):
        x = self.conv1(x, edge_index) + self.lin1(x)
        x = x.relu()
        x = self.conv2(x, edge_index) + self.lin2(x)
        return x


model = HeteroGNN(hidden_channels=4, out_channels=4)
hgtmodel = to_hetero(model, data.metadata(), aggr='sum')

hgtmodel.train()
out = hgtmodel(data.x_dict, data.edge_index_dict)
print(out)
```

**Reply by Zhang-1218** [reply]
？？？？
incredible? ? ?
I will try again

**Reply by Zhang-1218** [reply]
I copied your code exactly, but still getting the same error.
collapse...
Is it because of the pyg version?
Here is the version info I have installed:
```
pytorch                   1.8.0           py3.8_cuda10.1_cudnn7.6.3_0    pytorch
pytorch-cluster           1.5.9           py38_torch_1.8.0_cu101         pyg
pytorch-scatter           2.0.8           py38_torch_1.8.0_cu101         pyg
pytorch-sparse            0.6.12          py38_torch_1.8.0_cu101         pyg
pytorch-spline-conv       1.2.1           py38_torch_1.8.0_cu101         pyg
torch-geometric           2.0.2                    pypi_0                pypi
torchaudio                0.8.0                      py38                pytorch
torchvision               0.9.0                py38_cu101                pytorch
```

**Reply by rusty1s** [reply]
Oh, you are right. You will need to install PyG from master to get this fixed:
```
pip install git+https://github.com/pyg-team/pytorch_geometric.git
```
The reason is that previously, any names with a `-` in it were not parsed correctly. Alternatively, you can change the name of your edge type to `co_occurence`.

**Reply by Zhang-1218** [reply]
Omg, it's really a version issue
I'm so stupid, why use -, not _
Finally done.

I also have a small question, which is to define different convolution kernels for different edge types, so how should the attributes of the edge be passed in?
```
class HeteroGNN(torch.nn.Module):
    def __init__(self, hidden_channels, out_channels, num_layers):
        super().__init__()

        self.convs = torch.nn.ModuleList()
        for _ in range(num_layers):
            conv = HeteroConv({
                ('paper', 'cites', 'paper'): GCNonv(-1, hidden_channels),
                ('author', 'writes', 'paper'): SAGEConv((-1, -1), hidden_channels),
                ('paper', 'rev_writes', 'author'): GATConv((-1, -1), hidden_channels),
            }, aggr='sum')
            self.convs.append(conv)

        self.lin = Linear(hidden_channels, out_channels)

    def forward(self, x_dict, edge_index_dict):
        for conv in self.convs:
            x_dict = conv(x_dict, edge_index_dict)
            x_dict = {key: x.relu() for key, x in x_dict.items()}
        return self.lin(x_dict['author'])

model = HeteroGNN(hidden_channels=64, out_channels=dataset.num_classes,
                  num_layers=2)
```

In the above code, how to pass the one-dimensional edge_weight between paper and paper to GCN, and GAT can obtain multi-dimensional edge_attr?

**Reply by rusty1s** [reply]
```python
conv = HeteroConv({
    ('paper', 'cites', 'paper'): GCNonv(-1, hidden_channels),
    ('author', 'writes', 'paper'): SAGEConv((-1, -1), hidden_channels),
    ('paper', 'rev_writes', 'author'): GATConv((-1, -1), hidden_channels, edge_dim=...),
}, aggr='sum')

conv(data.x_dict, data.edge_index_dict, data.edge_attr_dict)
```
should work just fine. In case a layer can take `edge_attr` as input, it will take it from `data.edge_attr_dict`. Let me know if it works for you :)

**Reply by Zhang-1218** [reply]
Emmm
it doesn't work, GCNConv can't find edge_dim

```
 GCNConv(-1, hidden_channels, edge_dim=-1),
  File "/root/anaconda3/envs/lib/python3.8/site-packages/torch_geometric/nn/conv/gcn_conv.py", line 126, in __init__
    super(GCNConv, self).__init__(**kwargs)
TypeError: __init__() got an unexpected keyword argument 'edge_dim'
```

**Reply by rusty1s** [reply]
`GCNConv` does not have an `edge_dim` attribute. I'm sorry: I read `GATConv` instead of `GCNConv`. I fixed the example in my code snippet above.

**Reply by Zhang-1218** [reply]
Emm
I have tried modifying it, but the problem still exists.
```
conv = HeteroConv({
    ('paper', 'cites', 'paper'): GCNonv(-1, hidden_channels),
    ('author', 'writes', 'paper'): SAGEConv((-1, -1), hidden_channels),
    ('paper', 'rev_writes', 'author'): GATConv((-1, -1), hidden_channels, edge_dim=...),
}, aggr='sum')

ERROR:

GATConv((-1, -1), hidden_channels, edge_dim=...),
  File "/root/anaconda3/envs/lib/python3.8/site-packages/torch_geometric/nn/conv/gat_conv.py", line 147, in __init__
    self.lin_edge = Linear(edge_dim, heads * out_channels, bias=False,
  File "/root/anaconda3/envs/hang/lib/python3.8/site-packages/torch_geometric/nn/dense/linear.py", line 58, in __init__
    if in_channels > 0:
TypeError: '>' not supported between instances of 'ellipsis' and 'int'
```

And in this case, there is no way for GCN to learn the weights of the edges.?

**Reply by rusty1s** [reply]
If you have edge features for `('paper', 'cites', 'paper')` relation, why not make use of GNN layer that can support it? Note that you will need to replace the `edge_dim=...` attribute of `GATConv` with the actual value of `edge_dim`, in your case `edge_dim=3`.

**Reply by Zhang-1218** [reply]
oh, I'm so sorry, I forgot to reply to you
My problem has been solved, thank you very much for your answer
